File: packages/sdk/python/robloxforge/terrain.py
"""Terrain generation methods for the RobloxForge SDK."""

import time
from typing import TYPE_CHECKING, Optional, Literal

if TYPE_CHECKING:
    from .client import Client


class TerrainBuild:
    def __init__(self, data: dict):
        self.build_id: str = data["buildId"]
        self.status: str = data["status"]
        self.estimated_seconds: int = data.get("estimatedSeconds", 60)
        self.download_url: Optional[str] = data.get("downloadUrl")
        self.tokens_used: Optional[int] = data.get("tokensUsed")

    def __repr__(self):
        return f"<TerrainBuild id={self.build_id} status={self.status}>"


class TerrainClient:
    def __init__(self, client: "Client"):
        self._client = client

    def generate(
        self,
        prompt: str,
        style: Literal["realistic", "cartoon", "lowpoly"] = "realistic",
        size: int = 1024,
        seed: Optional[int] = None,
    ) -> TerrainBuild:
        """
        Generate a Roblox terrain map from a text prompt.

        Args:
            prompt: Natural language description of the terrain
            style: Art style (realistic, cartoon, lowpoly)
            size: Map size in studs (512-4096)
            seed: Optional random seed for reproducibility

        Returns:
            TerrainBuild with build ID and initial status
        """
        payload = {"prompt": prompt, "style": style, "size": size}
        if seed is not None:
            payload["seed"] = seed

        data = self._client.request("POST", "/api/ai/generate/terrain", payload)
        return TerrainBuild(data)

    def get(self, build_id: str) -> TerrainBuild:
        """Get status of an existing terrain build."""
        data = self._client.request("GET", f"/api/ai/generate/terrain/{build_id}")
        return TerrainBuild(data)

    def wait_for_build(
        self,
        build_id: str,
        poll_interval: float = 2.0,
        timeout: float = 120.0,
    ) -> TerrainBuild:
        """
        Poll until a build completes or fails.

        Args:
            build_id: Build ID from generate()
            poll_interval: Seconds between polls (default 2.0)
            timeout: Max wait time in seconds (default 120.0)

        Returns:
            Completed or failed TerrainBuild

        Raises:
            TimeoutError: If build doesn't complete within timeout
        """
        deadline = time.time() + timeout
        while time.time() < deadline:
            build = self.get(build_id)
            if build.status in ("complete", "failed"):
                return build
            time.sleep(poll_interval)
        raise TimeoutError(f"Build {build_id} timed out after {timeout}s")

File: packages/sdk/typescript/src/terrain.ts
import { RobloxForge } from './client'

export interface TerrainGenerateOptions {
  prompt: string
  style?: 'realistic' | 'cartoon' | 'lowpoly'
  size?: 512 | 1024 | 2048 | 4096
  seed?: number
}

export interface TerrainBuild {
  buildId: string
  status: 'queued' | 'processing' | 'complete' | 'failed'
  estimatedSeconds: number
  downloadUrl: string | null
  tokensUsed?: number
}

export class TerrainClient {
  constructor(private rf: RobloxForge) {}

  /**
   * Generate a terrain map from a natural language prompt.
   */
  async generate(opts: TerrainGenerateOptions): Promise<TerrainBuild> {
    const res = await this.rf.request<TerrainBuild>('POST', '/api/ai/generate/terrain', opts)
    if (res.error) throw new Error(res.error)
    return res.data!
  }

  /**
   * Poll build status until complete or failed.
   * @param buildId - Build ID from generate()
   * @param pollIntervalMs - How often to check (default 2000ms)
   * @param timeoutMs - Max wait time (default 120000ms)
   */
  async waitForBuild(
    buildId: string,
    pollIntervalMs = 2000,
    timeoutMs = 120_000
  ): Promise<TerrainBuild> {
    const deadline = Date.now() + timeoutMs

    while (Date.now() < deadline) {
      const res = await this.rf.request<TerrainBuild>('GET', `/api/ai/generate/terrain/${buildId}`)
      if (res.error) throw new Error(res.error)

      const build = res.data!
      if (build.status === 'complete' || build.status === 'failed') {
        return build
      }

      await new Promise((resolve) => setTimeout(resolve, pollIntervalMs))
    }

    throw new Error(`Build ${buildId} timed out after ${timeoutMs}ms`)
  }
}

/**
 * Attach terrain methods to a RobloxForge instance.
 */
export function terrainPlugin(rf: RobloxForge) {
  return new TerrainClient(rf)
}
